File: msr/reference/typescript/balanced-parens.ts
function is_balanced(s: string): boolean {
  const closer: Record<string, string> = { "(": ")", "[": "]", "{": "}" };
  const stack: string[] = [];
  for (const c of s) {
    if (c in closer) stack.push(closer[c]);
    else if (c === ")" || c === "]" || c === "}") {
      if (stack.pop() !== c) return false;
    }
  }
  return stack.length === 0;
}

File: msr/reference/rust/balanced-parens.rs
fn is_balanced(s: &str) -> bool {
    let mut stack = Vec::new();
    for c in s.chars() {
        match c {
            '(' => stack.push(')'),
            '[' => stack.push(']'),
            '{' => stack.push('}'),
            ')' | ']' | '}' => {
                if stack.pop() != Some(c) {
                    return false;
                }
            }
            _ => {}
        }
    }
    stack.is_empty()
}
